src/smtencoding/FStar.SMTEncoding.Encode.fst

```diff
@@ -1000,6 +1000,547 @@ let encode_top_level_let :
+      | _ ->
+        let ttok_decl = Term.DeclFun(ttok, [], Term_sort, Some "token") in
+        let ttok_fresh = Term.fresh_token (ttok, Term_sort) (varops.next_id()) in
+        let ttok_app = mk_Apply ttok_tm vars in
+        let pats = [[ttok_app]; [tapp]] in
+        // These patterns allow rewriting (ApplyT T@tok args) to (T args) and vice versa
+        // This seems necessary for some proofs, but the bidirectional rewriting may be inefficient
+        let name_tok_corr =
+          Util.mkAssume(mkForall' (Ident.range_of_lid t) (pats, None, vars, mkEq(ttok_app, tapp)),
+                        Some "name-token correspondence",
+                        ("token_correspondence_"^ttok)) in
+        [ttok_decl; ttok_fresh; name_tok_corr], env
+    in
+    tname_decl@tok_decls, env
+  in
+  let kindingAx =
+    let k, decls = encode_term_pred None res env' tapp in
+    let karr =
+      if List.length formals > 0
+      then [Util.mkAssume(mk_tester "Tm_arrow" (mk_PreType ttok_tm), Some "kinding", ("pre_kinding_"^ttok))]
+      else []
+    in
+    let rng = Ident.range_of_lid t in
+    let tot_fun_axioms = EncodeTerm.isTotFun_axioms rng ttok_tm vars (List.map (fun _ -> mkTrue) vars) true in
+    decls@(karr@[Util.mkAssume(mkAnd(tot_fun_axioms, mkForall rng ([[tapp]], vars, mkImp(guard, k))),
+                                None,
+                                ("kinding_"^ttok))] |> mk_decls_trivial)
+  in
+  let aux =
+    kindingAx
+    @(inversion_axioms env tapp vars)
+    @([pretype_axiom (Ident.range_of_lid t) env tapp vars] |> mk_decls_trivial)
+  in
+  (decls |> mk_decls_trivial)@binder_decls@aux, env
+
+let encode_datacon (is_injective_on_tparams:bool) (env:env_t) (se:sigelt)
+: decls_t * env_t
+= let Sig_datacon {lid=d; t; num_ty_params=n_tps; mutuals} = se.sigel in
+  let quals = se.sigquals in
+  let t = norm_before_encoding env t in
+  let formals, t_res = U.arrow_formals t in
+  let arity = List.length formals in
+  let ddconstrsym, ddtok, env = new_term_constant_and_tok_from_lid env d arity in
+  let ddtok_tm = mkApp(ddtok, []) in
+  let fuel_var, fuel_tm = fresh_fvar env.current_module_name "f" Fuel_sort in
+  let s_fuel_tm = mkApp("SFuel", [fuel_tm]) in
+  let vars, guards, env', binder_decls, names = encode_binders (Some fuel_tm) formals env in
+  let is_injective_on_tparams =
+    is_injective_on_tparams || Options.ext_getv "compat:injectivity" <> ""
+  in
+  let fields =
+    names |>
+    List.mapi
+      (fun n x ->
+        let field_projectible =
+          n >= n_tps || //either this field is not a type parameter
+          is_injective_on_tparams //or we are allowed to be injective on parameters
+        in
+        { field_name=mk_term_projector_name d x;
+          field_sort=Term_sort;
+          field_projectible })
+  in
+  let datacons = {
+    constr_name=ddconstrsym;
+    constr_fields=fields;
+    constr_sort=Term_sort;
+    constr_id=Some (varops.next_id());
+    constr_base=not is_injective_on_tparams
```

Comment:
Generate the base constructor for irrelevance of type param arguments, if not injective on type params